src/Elasticsearch/Endpoints/AbstractEndpoint.php

```diff
@@ -60,14 +64,18 @@
     /**
      * @param Transport $transport
      */
-    public function __construct($transport)
+    public function __construct($transport, RequestFactory $requestFactory = null)
     {
         $this->transport = $transport;
+        $this->requestFactory = $requestFactory;
     }
 
     /**
      * @throws \Exception
      * @return array
+     *
+     * @deprecated Performing a request directly on an endpoint is deprecated since 2.2 and will be removed in 3.0,
+     *             you should instead use the createRequest method
      */
     public function performRequest()
```

Comment:
> Also the performRequest use an options parameters to set specific options to the underlying http client, this is not possible anymore.
> AFAIK it was only used for verbose in some parts, but don't know the usage it was done by users ?

Oh, there's no way to send custom parameters to the underlying HTTP implementation?  That's sorta a deal breaker?

I'm looking at the change closer now, and seem some issues.  I'm not familiar with PSR7 enough to say if this is a problem or not, but here are some potential snags:

- We need signal to the client that a request should be run in `async` mode, this is done through setting `['client']['future']`.  This isn't technically passed to RingPHP, but is part of the endpoint's `$options` map.  Same for `['client']['ignore']`, which tells the client to ignore certain classes of HTTP errors so an exception is not thrown

- Currently you can set custom curl options/headers through `['client']['curl']`.  I believe PSR7 allows setting this kind of information, but the translation will need to be done

- Verbosity can be toggled via `['client']['verbose']` as you mentioned.  This is used in a few places internally, and is a very useful debugging tool for the end-user.  It instructs RingPHP to emit all the info from `CURLOPT_VERBOSE` flag.  There's likely an equivalent in PSR-7 land, but it'd need translation for sure.

- Variety of other things, `['client']['never_retry']`, `['client']['timeout']`, etc

Full disclosure, I'm dubious of the usefulness of converting the internals to PSR7.  I'm not sure what it really gains us?  In my eyes I think it makes more sense to have:

- All endpoints can return a PSR-7 compliant request object instead of executing it.  This allows the user to do whatever they want to the request, execute it with a different client entirely, etc.

- An endpoint that blindly executes PSR-7 compliant request objects.  This allows ES-PHP to execute any arbitrary HTTP request the user builds (including ones that were generated by the client but subsequently modified), and still get to leverage the ES-specific features like sniffing.

- Responses should probably be PSR-7 compliant, so that they can fit into anyone's PSR-7 pipeline.  Relatively inoculous.

So in light of that, perhaps it makes more sense to just write shims that can convert to/from PSR-7, instead of rewiring all the guts?  I'm just not sure I see much added benefit.  I'm fairly happy with RingPHP and don't really see the need to support umpteen different (nearly identical) HTTP libs.  They all either fallback to streams or libcurl and we don't need many of the features that full Guzzle provides, etc.

Happy to be convinced otherwise, but that's my current feelings :)  Perhaps I just need to read into PSR-7 some more, it's been a while since I last looked at it.